animal-starter/src/main/java/ru/mts/services/AnimalsRepositoryImpl.java

```diff
@@ -0,0 +1,120 @@
+        return animalsList.stream()
+                .filter(animal -> animal.getBirthDate().isLeapYear())
+                .collect(Collectors.toMap(animal -> animal.getClass().getSimpleName() + " " + animal.getName(), AbstractAnimal::getBirthDate));
+    }
+
+    @Override
+    public Map<AbstractAnimal, Integer> findOlderAnimal(int n) {
+        Map<AbstractAnimal, Integer> result = animalsList.stream()
+                .filter(animal -> getAnimalAge(animal) > n)
+                .collect(Collectors.toMap(entry -> entry, this::getAnimalAge));
+
+        if (!result.isEmpty()) {
+            return result;
+        }
+
+        AbstractAnimal oldestAnimal = animalsList.stream()
+                .sorted(Comparator.comparing(AbstractAnimal::getBirthDate))
+                .toList().get(0);
+        return Map.of(oldestAnimal, getAnimalAge(oldestAnimal));
+    }
+
+    @Override
+    public Map<String, List<AbstractAnimal>> findDuplicate() {
+        return animalsList.stream()
+                .collect(Collectors.groupingBy(Function.identity(),
+                        Collectors.counting()))
+                .entrySet().stream()
+                .filter(entry -> entry.getValue() > 1)
+                .map(entry -> Map.entry(entry.getKey().getClass().getSimpleName(), entry.getKey()))
+                .collect(Collectors.groupingBy(Map.Entry::getKey, Collectors.mapping(Map.Entry::getValue, Collectors.toList())));
+    }
+
+    @Override
+    public void printDuplicate() {
+        Map<String, List<AbstractAnimal>> duplicates = findDuplicate();
+        if (duplicates.isEmpty()) {
+            System.out.println("No duplicates");
+        }
+
+        for (Map.Entry<String, List<AbstractAnimal>> entry : duplicates.entrySet()) {
+            System.out.println(entry.getKey());
+            for (AbstractAnimal animal : entry.getValue()) {
+                System.out.println("   - " + animal);
+            }
+        }
+    }
+
+    @Override
+    public double findAverageAge() {
+        return animalsList.stream()
+                .mapToDouble(this::getAnimalAge)
+                .average()
+                .orElse(-1);
+    }
+
+    @Override
+    public List<AbstractAnimal> findOldAndExpensive() {
+        double avgCost = animalsList.stream()
+                .mapToDouble(animal -> animal.getCost().intValue())
+                .average()
+                .orElse(-1);
+        return animalsList.stream()
+                .filter(animal -> getAnimalAge(animal) > 5 && animal.getCost().doubleValue() > avgCost)
+                .sorted(Comparator.comparing(AbstractAnimal::getBirthDate))
+                .collect(Collectors.toList());
+    }
+
+    @Override
+    public List<AbstractAnimal> findMinCostAnimals() {
+        return animalsList.stream()
+                .sorted(Comparator.comparing(AbstractAnimal::getCost))
+                .limit(3)
+                .sorted((first, second) -> -first.getName().compareTo(second.getName()))
```

Comment:
умножение на -1 выглядит не оч (вкусовщина)

```suggestion
                .sorted(Comparator.comparing(AbstractAnimal::getName).reversed())
```